### PR TITLE
Add dt plots

### PR DESCRIPTION
**Pull request by Piphi5**
- Adds diagnostic plots to see the timing of data collection
- Stores test start positions for more robust identification of times
between tests

Example graph:
image

### CODE REVIEW

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 1]
```diff
@@ -127,6 +132,17 @@ void AnalyzerPlot::SetData(const Storage& data, const std::vector<double>& ff,
...
     m_data[kChartTitles[3]].emplace_back(slow[i].timestamp - t,
                                          slow[i].acceleration);
+
+    if (i > 0) {
+      // If the current timestamp is in the startTimes array, it is the
+      // beginning of a new test and the dt will be inflated. Therefore we skip
+      // it to exclude that dt and effectively reset dt calculations.
+      if (std::find(startTimes.begin(), startTimes.end(), slow[i].timestamp) ==
+          startTimes.end()) {
+        double dt = (slow[i].timestamp - slow[i - 1].timestamp) * 1000;
```
Comment (calcmogul) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:142]: If `i` is the start time, subtracting the timestamp at `i - 1` will result in a large dt.

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 2]
```diff
@@ -152,24 +168,36 @@ void AnalyzerPlot::SetData(const Storage& data, const std::vector<double>& ff,
                                          fast[i].velocity);
     m_data[kChartTitles[5]].emplace_back(fast[i].timestamp - t,
                                          fast[i].acceleration);
+    if (i > 0) {
+      // If the current timestamp is in the startTimes array, it is the
+      // beginning of a new test and the dt will be inflated. Therefore we skip
+      // it to exclude that dt and effectively reset dt calculations.
+      if (std::find(startTimes.begin(), startTimes.end(), fast[i].timestamp) ==
+          startTimes.end()) {
+        double dt = (fast[i].timestamp - fast[i - 1].timestamp) * 1000;
```
Comment (calcmogul) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:177]: Same comments as earlier.

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 3]
```diff
@@ -278,4 +306,22 @@ void AnalyzerPlot::DisplayTimeDomainPlots() {
       }
     }
   }
+  if (m_fitNextPlot[6]) {
+    ImPlot::FitNextPlotAxes();
+  }
```
Comment (calcmogul) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:311]: We should clamp the y minimum to 0 since negative periods are impossible.

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Comment by calcmogul** [comment]
It's not clear from the screenshot what the nomimal dt is supposed to be. We could draw a flat line at the mean perhaps.

File: src/main/native/cpp/view/AnalyzerPlot.cpp (hunk 1 above)
Comment (Piphi5) [reply src/main/native/cpp/view/AnalyzerPlot.cpp:142]: Doesn't the if statement skip over values where the timestamp at `i` is a start time?

File: src/main/native/cpp/view/AnalyzerPlot.cpp (hunk 3 above)
Comment (Piphi5) [reply src/main/native/cpp/view/AnalyzerPlot.cpp:311]: Do I set the plot limits or should I remove any negative datapoints?

**Comment by Piphi5** [comment]
@calcmogul does this work?
image

**Comment by prateekma** [comment]
Instead of autoscaling it, we should set the limits on this one because most dts will always be between 0 and 25.

**Comment by Piphi5** [comment]
@prateekma 
is this sizing (0-50ms) good?
image

**Comment by prateekma** [comment]
Yes

**Review by calcmogul — APPROVED** [review]

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 4]
```diff
@@ -278,4 +327,27 @@ void AnalyzerPlot::DisplayTimeDomainPlots() {
       }
     }
   }
+  if (m_fitNextPlot[6]) {
+    ImPlot::SetNextPlotLimits(0, m_dtMaxTime, 0, 50);
```
Comment (prateekma) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:331]: Can we add a fit call just for the x-axis after this? See the arguments here: https://github.com/epezent/implot/blob/master/implot.h#L522

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 5]
```diff
@@ -278,4 +327,28 @@ void AnalyzerPlot::DisplayTimeDomainPlots() {
...
+    ImPlot::FitNextPlotAxes(true, false, false, false);
```
Comment (Piphi5) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:332]: @prateekma  is this what you were asking for?

Comment (prateekma) [reply src/main/native/cpp/view/AnalyzerPlot.cpp:332]: Yes

Comment (Piphi5) [reply src/main/native/cpp/view/AnalyzerPlot.cpp:332]: Is it good to merge then?

Comment (prateekma) [reply src/main/native/cpp/view/AnalyzerPlot.cpp:332]: Yes, I'll review again tonight to make sure but it should be good to go.

**Review by prateekma — APPROVED** [review]

**Pull request merged**